compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1692,10 +1692,10 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     param_env = self.tcx.param_env(unevaluated.def.did);
                 }
                 Ok(Some(ct)) => {
-                    if ct.unify_failure_kind(self.tcx) == FailureKind::Concrete {
-                        substs = replace_param_and_infer_substs_with_placeholder(self.tcx, substs);
-                    } else {
+                    if ct.has_infer_types_or_consts() || ct.has_param_types_or_consts() {
                         return Err(ErrorHandled::TooGeneric);
+                    } else {
+                        substs = replace_param_and_infer_substs_with_placeholder(self.tcx, substs);
```

Comment:
I think I changed this because now the const may not be an `Expr`, so it may not necessarily have `unify_failure_kind` defined. Since we substitute params immediately though, we can just directly check if there are any parameters/inference variables.